stackslib/src/chainstate/nakamoto/staging_blocks.rs

```diff
@@ -407,6 +464,114 @@ impl<'a> NakamotoStagingBlocksTx<'a> {
+        block: &NakamotoBlock,
+        burn_attachable: bool,
+        signing_weight: u32,
+        obtain_method: NakamotoBlockObtainMethod,
+    ) -> Result<(), ChainstateError> {
+        let Ok(tenure_start) = block.is_wellformed_tenure_start_block() else {
+            return Err(ChainstateError::InvalidStacksBlock(
+                "Tried to store a tenure-start block that is not well-formed".into(),
+            ));
+        };
+
+        let burn_attachable = if !burn_attachable {
+            // if it's burn_attachable before, it's burn_attachable always
+            self.conn()
+                .is_burn_block_processed(&block.header.consensus_hash)?
+        } else {
+            burn_attachable
+        };
+
+        self.execute(
+            "INSERT INTO nakamoto_staging_blocks (
+                     block_hash,
+                     consensus_hash,
+                     parent_block_id,
+                     is_tenure_start,
+                     burn_attachable,
+                     orphaned,
+                     processed,
+
+                     height,
+                     index_block_hash,
+                     processed_time,
+                     obtain_method,
+                     signing_weight,
+                     
+                     data
+            ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13)",
+            params![
+                &block.header.block_hash(),
+                &block.header.consensus_hash,
+                &block.header.parent_block_id,
+                &tenure_start,
+                if burn_attachable { 1 } else { 0 },
+                0,
+                0,
+                u64_to_sql(block.header.chain_length)?,
+                &block.block_id(),
+                0,
+                obtain_method.to_string(),
+                signing_weight,
+                block.serialize_to_vec(),
+            ],
+        )?;
+        if burn_attachable {
+            self.set_burn_block_processed(&block.header.consensus_hash)?;
+        }
+        Ok(())
+    }
+
+    /// Do we have a block with the given sighash?
+    /// NOTE: the block hash and sighash are the same for Nakamoto blocks
+    pub(crate) fn has_nakamoto_block_with_sighash(
+        &self,
+        consensus_hash: &ConsensusHash,
+        sighash: &Sha512Trunc256Sum,
+    ) -> Result<bool, ChainstateError> {
+        let qry =
+            "SELECT 1 FROM nakamoto_staging_blocks WHERE consensus_hash = ?1 AND block_hash = ?2";
+        let args = rusqlite::params![consensus_hash, sighash];
+        let present: Option<u32> = query_row(self, qry, args)?;
+        Ok(present.is_some())
+    }
+
+    /// Store a block into the staging DB if its sighash has never been seen before.
+    /// NOTE: the block hash and sighash are the same for Nakamoto blocks, so this is equivalent to
+    /// storing a new block.
+    /// Return true if stored; false if not.
+    pub(crate) fn try_store_block_with_new_sighash(
```

Comment:
Again, I think this would be an abstraction leak.  The goal of these methods is to allow the caller to see if it can store (and later process) a malleablized block.  Malleablized blocks all have the same signer signature hash, since malleablization only occurs through varying the signature vector.  The fact that we also happen to treat the signer signature hash as the block hash is an implementation detail.

Reply:
I don't think it'd be an abstraction leak though -- the caller has the block, so they can compute the block hash and stacks block id of the block. The caller actually doesn't need to reason about the sighash at all in this case. If that were to change in the future, the argument types would change, and the caller would have to be updated to pass type checks.